pkg/framework/resource/parse.go

```diff
@@ -0,0 +1,211 @@
+	return &parser{
+		resources,
+		discovery,
+		apiGroup,
+		apiVersion,
+	}
+}
+
+// Resources discovers and indexes resources from the API server.
+// It returns a Resources object: a map of resource names and their associated Resource objects, ordered by preference as reported by the server.
+func (p *parser) Resources() (Resources, error) {
+	gvs, err := p.discovery.ServerResources()
+	if err != nil {
+		return nil, err
+	}
+	resources, byGVR := p.indexResources(gvs)
+	err = p.attachSubResources(gvs, resources, byGVR)
+	return resources, err
+}
+
+// subResource returns a resource name, subresource name pair, and true if the resource is a subresource.
+// It returns a resource name, an empty string, and false if the resource is not a subresource.
+func (*parser) subResource(resource *v1.APIResource) (string, string, bool) {
+	parts := strings.Split(resource.Name, "/")
+	if len(parts) > 1 {
+		return parts[0], parts[1], true
+	}
+	return parts[0], "", false
+}
+
+// resource returns the resource name and true if it is a resource (not a subresource).
+// It returns the resource name and false it if it is a subresource.
+func (p *parser) resource(resource *v1.APIResource) (string, bool) {
+	r, _, b := p.subResource(resource)
+	return r, !b
+}
+
+// copyGroupVersion copies the group and version from source to destination if either is missing on the destination.
+// If the source group is empty and the destination group is empty, sets the destination group to "core".
+func (*parser) copyGroupVersion(src *v1.APIResource, dest *v1.APIResource) {
+	if len(dest.Group) == 0 {
+		dest.Group = src.Group
+	}
+	if len(dest.Group) == 0 {
+		dest.Group = "core"
+	}
+	if len(dest.Version) == 0 {
+		dest.Version = src.Version
+	}
+}
+
+// splitGroupVersion splits the groupVersion string into its group and version components.
+func (*parser) splitGroupVersion(groupVersion string) (string, string) {
+	parts := strings.Split(groupVersion, "/")
+	var group, version string
+	if len(parts) > 1 {
+		group = parts[0]
+	}
+	if len(parts) > 1 {
+		version = parts[1]
+	} else if len(parts) > 0 {
+		version = parts[0]
+	} else {
+		version = "v1"
+	}
+	return group, version
+}
+
+// defaultGroupVersion sets the group and version to the API group and version if they're missing.
+func (p *parser) defaultGroupVersion(resource *v1.APIResource, group, version string) {
+	if len(resource.Group) == 0 {
+		resource.Group = group
+	}
+	if len(resource.Version) == 0 {
+		resource.Group = version
+	}
+}
+
+// isGroupVersionMatch returns false if either group or version doesn't match with the API.
+func (p *parser) isGroupVersionMatch(group, version string) bool {
+	if len(p.apiGroup) > 0 && p.apiGroup != group {
+		return false
+	}
+	if len(p.apiVersion) > 0 && p.apiVersion != version {
+		return false
+	}
+	return true
+}
+
+// getOpenAPI retrieves a schema object from the API based on a GroupVersionResource triplet.
```

Comment:
I'm not sure why this returns the Schema and bool. It's safe and sufficient to check if Schema is nil.